### PR TITLE
Upgrade to Manta v1.0.0

### LINKED ISSUE #811: Fix MantaPay v1 tests
Most of the tests need to be updated with changes from https://github.com/Manta-Network/Manta/pull/748

### LINKED ISSUE #812: Integrate MantaPay V1 changes into production/mock runtimes

### PR DESCRIPTION
**Pull request by bhgomes**
Signed-off-by: Brandon H. Gomes <bhgomes@pm.me>

## Description

relates to OR closes: #XXXX

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Added **one** label out of the `L-` group to this PR
- [ ] Added **one or more** labels out of `A-` and `C-` groups to this PR
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

Situational Notes:
- If adding functionality, write unit tests!
- If importing a new pallet, choose a proper module index for it, and allow it in `BaseFilter`. Ensure **every** extrinsic works from front-end. If there's corresponding tool, ensure both work for each other.
- If needed, update our Javascript/Typescript APIs. These APIs are officially used by exchanges or community developers.
- If modifying existing runtime storage items, make sure to implement storage migrations for the runtime and test them with `try-runtime`. This includes migrations inherited from upstream changes, and you can search the diffs for modifications of `#[pallet::storage]` items to check for any.
- If runtime changes, need to update the version numbers properly:
   * `authoring_version`: The version of the authorship interface. An authoring node will not attempt to author blocks unless this is equal to its native runtime.
   * `spec_version`: The version of the runtime specification. A full node will not attempt to use its native runtime in substitute for the on-chain Wasm runtime unless all of spec_name, spec_version, and authoring_version are the same between Wasm and native.
   * `impl_version`: The version of the implementation of the specification. Nodes are free to ignore this; it serves only as an indication that the code is different; as long as the other two versions are the same then while the actual code may be different, it is nonetheless required to do the same thing. Non-consensus-breaking optimizations are about the only changes that could be made which would result in only the impl_version changing.
   * `transaction_version`: The version of the extrinsics interface. This number must be updated in the following circumstances: extrinsic parameters (number, order, or types) have been changed; extrinsics or pallets have been removed; or the pallet order in the construct_runtime! macro or extrinsic order in a pallet has been changed. You can run the `metadata_diff.yml` workflow for help. If this number is updated, then the `spec_version` must also be updated
- Verify benchmarks & weights have been updated for any modified runtime logics

### CODE REVIEW

File: primitives/manta/src/assets.rs [hunk 1]
```diff
@@ -306,210 +367,267 @@ impl FungibleLedgerError {
...
+    ) -> Result<(), FungibleLedgerError<Self::AssetId, Self::Balance>>;
+
+    /// Checks if `amount` of `asset_id` can be deposited into `account` and then deposits it if
+    /// that is successful.
+    fn try_deposit_minting(
+        asset_id: Self::AssetId,
+        account: &Self::AccountId,
+        amount: Self::Balance,
+        can_increase_total_supply: bool,
+    ) -> Result<(), FungibleLedgerError<Self::AssetId, Self::Balance>>;
```
Comment (ghzlatarev) [inline primitives/manta/src/assets.rs:431]: +1 for the nicer errors.

File: primitives/manta/src/assets.rs [hunk 2]
```diff
@@ -306,210 +367,267 @@ impl FungibleLedgerError {
...
-        amount: Balance,
+        asset_id: Self::AssetId,
+        source: &Self::AccountId,
+        destination: &Self::AccountId,
+        amount: Self::Balance,
+        existence_requirement: ExistenceRequirement,
+    ) -> Result<(), FungibleLedgerError<Self::AssetId, Self::Balance>>;
+
+    /// Check whether `account` can decrease its balance by `amount` in the given `asset_id`.
+    fn can_withdraw(
```
Comment (ghzlatarev) [inline primitives/manta/src/assets.rs:443]: nice

File: primitives/manta/src/assets.rs [hunk 3]
```diff
@@ -144,48 +216,35 @@ impl Default for AssetRegistrarMetadata {
...
+    type Balance = B;
+}
+
+impl<B> AssetMetadata for AssetRegistryMetadata<B> {
+    #[inline]
+    fn min_balance(&self) -> &B {
+        &self.min_balance
     }
 
+    #[inline]
```
Comment (ghzlatarev) [inline primitives/manta/src/assets.rs:236]: I'm going to leave this unresolved comment for myself, to run the benchmarks with/without all the inlines. Curious how much it actually changes the perf.

File: pallets/manta-pay/src/types.rs [hunk 4]
```diff
@@ -50,25 +59,60 @@ where
     T::decode(&mut bytes.as_slice())
 }
 
+///
```
Comment (ghzlatarev) [inline pallets/manta-pay/src/types.rs:62]: missing comments.

File: primitives/manta/src/xcm.rs [hunk 5]
```diff
@@ -224,78 +229,68 @@ impl<
...
 
+    ///
+    #[inline]
     fn refund_weight(&mut self, weight: Weight) -> Option<MultiAsset> {
-        if let Some((id, prev_amount, units_per_second)) = self.refund_cache.clone() {
+        if let Some((id, prev_amount, units_per_second)) = &mut self.refund_cache {
             let weight = weight.min(self.weight);
             self.weight -= weight;
-            let amount = units_per_second * (weight as u128) / (WEIGHT_PER_SECOND as u128);
-            self.refund_cache = Some((
```
Comment (ghzlatarev) [inline primitives/manta/src/xcm.rs:241]: nice refactor

File: primitives/manta/src/assets.rs [hunk 6]
```diff
@@ -306,210 +367,267 @@ impl FungibleLedgerError {
     }
 }
 
+impl<I, B> AssetIdType for FungibleLedgerError<I, B> {
```
Comment (ghzlatarev) [inline primitives/manta/src/assets.rs:386]: ooh ok , so this was how to make these two generic...

File: pallets/manta-pay/src/benchmark/mod.rs [hunk 7]
```diff
@@ -127,17 +128,18 @@ benchmarks! {
...
         sink.clone()
     ) verify {
         // FIXME: add balance checking
-        assert_last_event::<T, _>(Event::Transfer { asset, source: caller.clone(), sink, });
+        assert_last_event::<T, _>(Event::Transfer { asset, source: caller.clone(), sink });
     }
 }
 
-impl_benchmark_test_suite!(Pallet, crate::mock::new_test_ext(), crate::mock::Test);
+// TODO: impl_benchmark_test_suite!(Pallet, crate::mock::new_test_ext(), crate::mock::Test);
```
Comment (ghzlatarev) [inline pallets/manta-pay/src/benchmark/mod.rs:148]: todo

File: pallets/manta-pay/src/benchmark/mod.rs [hunk 8]
```diff
@@ -90,26 +90,27 @@ benchmarks! {
...
         // FIXME: add balance checking
-        assert_last_event::<T, _>(Event::ToPublic { asset: Asset::new(8, 10_000), sink: caller });
+        assert_last_event::<T, _>(Event::ToPublic { asset, sink: caller });
     }
 
     private_transfer {
         let caller: T::AccountId = whitelisted_caller();
         let origin = T::Origin::from(RawOrigin::Signed(caller.clone()));
-        init_asset::<T>(&caller, 8u32, 1_000_000u128);
+        init_asset::<T>(&caller, Pallet::<T>::field_from_id(8), 1_000_000u128);
```
Comment (ghzlatarev) [inline pallets/manta-pay/src/benchmark/mod.rs:116]: We should probably fill up the ledger more for these benchmarks.

File: pallets/manta-pay/src/lib.rs [hunk 9]
```diff
@@ -834,101 +902,101 @@ where
...
                 &Pallet::<T>::account_id(),
-                withdraw.0,
+                withdraw,
                 ExistenceRequirement::KeepAlive,
             )?;
         }
         for WrapPair(account_id, deposit) in sinks {
             FungibleLedger::<T>::transfer(
-                asset_id.0,
+                Pallet::<T>::id_from_field(encode(asset_id)).expect("FIXME"),
```
Comment (ghzlatarev) [inline pallets/manta-pay/src/lib.rs:1012]: There are a bunch of `expect`s in this file. If we leave some we should have a comment why it's ok.
Same of the `unwrap`s.

File: pallets/manta-pay/src/lib.rs [hunk 10]
```diff
@@ -589,12 +622,36 @@ pub mod pallet {
...
+                        sources,
+                        sinks,
+                    )
                     .map_err(Error::<T>::from)?
                     .convert(origin),
             );
             Ok(().into())
         }
+
+        ///
```
Comment (ghzlatarev) [inline pallets/manta-pay/src/lib.rs:650]: Documentation

File: pallets/manta-pay/src/lib.rs [hunk 11]
```diff
@@ -589,12 +622,36 @@ pub mod pallet {
...
+            if u128::from_le_bytes(Array::from_iter(id[0..16].iter().copied()).into()) == 0 {
+                Some(u128::from_le_bytes(
+                    Array::from_iter(id[16..32].iter().copied()).into(),
+                ))
+            } else {
+                None
+            }
+        }
+
+        ///
```
Comment (ghzlatarev) [inline pallets/manta-pay/src/lib.rs:662]: Documentation

File: pallets/manta-pay/src/types.rs (hunk 4 above)
Comment (ghzlatarev) [inline pallets/manta-pay/src/types.rs:62]: bunch of missing docs on this page

**Comment by Dengjianping** [comment]
Can the rpc method `mantaPay_pull_ledger_diff` be renamed to `mantaPay_pullLedgerDiff`?
https://github.com/Manta-Network/Manta/pull/748/files#diff-ada1b74d9d0e41a7bc44d9d3ef17a29cb751ef83d034424adcb6c3a671f6c1a5L40

Because it doesn't follow camel case convention.

**Comment by Dengjianping** [comment]
And please address this issue: https://github.com/Manta-Network/Manta/issues/773

**Comment by ghzlatarev** [comment]
Continues with https://github.com/Manta-Network/Manta/pull/903

**Pull request closed without merging**